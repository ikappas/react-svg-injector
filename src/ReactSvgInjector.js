import React from "react";
import PropTypes from "prop-types";

const isBrowser = typeof window !== "undefined";
const SVGInjector = isBrowser ? require("svg-injector") : undefined;
const SvgContext = React.createContext();

class ReactSvgInjector extends React.Component {
  static defaultProps = {
    evalScripts: "once",
    callback: () => {}
  };
  static propTypes = {
    src: PropTypes.string,
    callback: PropTypes.func,
    evalScripts: PropTypes.oneOf(["always", "once", "never"])
  };
  state = {
    svg: null
  };
  componentDidMount() {
    const { evalScripts } = this.props;
    SVGInjector(this.img.current, {
      evalScripts,
      each: this.callback
    });
  }
  img = React.createRef();
  render() {
    const { src, evalScripts, callback, children, ...props } = this.props;
    return (
      <SvgContext.Provider value={this.state}>
        <img data-src={src} ref={this.img} {...props} />
        {children}
      </SvgContext.Provider>
    );
  }
  callback = svg => {
    this.setState({ svg });
    this.props.callback(svg);
  };
}

export { SvgContext, ReactSvgInjector };
