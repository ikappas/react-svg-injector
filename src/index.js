export { ReactSvgInjector } from "./ReactSvgInjector";
export { Mutate } from "./Mutate";
