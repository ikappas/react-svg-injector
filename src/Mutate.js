import React, { Component } from "react";
import PropTypes from "prop-types";
import { SvgContext } from "./ReactSvgInjector";

class Mutate extends Component {
  static propTypes = {
    selector: PropTypes.string.isRequired
  };
  render() {
    return (
      <SvgContext.Consumer>
        {state => {
          if (state.svg) {
            const { selector, ...props } = this.props;
            const nodeList = state.svg.querySelectorAll(selector);
            for (let node of nodeList.values()) {
              for (let [key, value] of Object.entries(props)) {
                if (typeof value === "function") {
                  // NOTE: covers mose cases, like onClick however I haven't tested
                  // if any other DOM elements take functions that aren't on* events
                  node[key.toLowerCase()] = value;
                } else {
                  node.setAttributeNS(null, key, value);
                }
              }
            }
          }
          return null;
        }}
      </SvgContext.Consumer>
    );
  }
}

export { Mutate };
