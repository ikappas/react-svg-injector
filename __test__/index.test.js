import { renderIntoDocument, wait } from "react-testing-library";
import React from "react";
import sinon from "sinon";
import { Mutate } from "../src/Mutate";
import { ReactSvgInjector } from "../src/ReactSvgInjector";
import * as src from "../src";

describe("index", () => {
  let container;
  let xhr;
  let requests = [];

  beforeEach(() => {
    container = document.body.appendChild(document.createElement("div"));
    xhr = sinon.useFakeXMLHttpRequest();
    xhr.onCreate = xhr => {
      requests.push(xhr);
    };
  });

  afterEach(() => {
    requests = [];
    xhr.restore();
    document.body.removeChild(container);
    jest.clearAllTimers();
  });

  describe("exports", () => {
    it("is expected to export ReactSvgInjector and Mutate", () => {
      expect(src.ReactSvgInjector).toBeDefined();
      expect(src.Mutate).toBeDefined();
    });
  });

  // @TODO: until this is released, we can't actually test the API changes: https://github.com/airbnb/enzyme/pull/1513
  describe("integration", () => {
    it("renders an SVG", async () => {
      const { container, debug } = renderIntoDocument(
        <ReactSvgInjector src="../__fixtures__/logo.svg">
          <Mutate fill="red" selector="g" />
        </ReactSvgInjector>
      );
      expect(container).toMatchSnapshot();
    });
  });
});
