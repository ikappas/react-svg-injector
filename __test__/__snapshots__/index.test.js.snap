// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index integration renders an SVG 1`] = `
<div>
  <img
    data-src="../__fixtures__/logo.svg"
    src=""
  />
</div>
`;
